chainstate/src/detail/transaction_verifier/flush.rs

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::{
+    storage::{TransactionVerifierStorageError, TransactionVerifierStorageMut},
+    token_issuance_cache::{CachedAuxDataOp, CachedTokenIndexOp, ConsumedTokenIssuanceCache},
+    CachedInputsOperation, TransactionVerifierDelta,
+};
+use common::chain::OutPointSourceId;
+
+fn flush_tx_indexes(
+    storage: &mut impl TransactionVerifierStorageMut,
+    tx_id: OutPointSourceId,
+    tx_index_op: CachedInputsOperation,
+) -> Result<(), TransactionVerifierStorageError> {
+    match tx_index_op {
+        CachedInputsOperation::Write(ref tx_index) => {
+            storage.set_mainchain_tx_index(&tx_id, tx_index)?
+        }
+        CachedInputsOperation::Read(_) => (),
+        CachedInputsOperation::Erase => storage.del_mainchain_tx_index(&tx_id)?,
+    }
+    Ok(())
+}
+
+fn flush_tokens(
+    storage: &mut impl TransactionVerifierStorageMut,
+    token_cache: &ConsumedTokenIssuanceCache,
+) -> Result<(), TransactionVerifierStorageError> {
+    assert_eq!(token_cache.data.len(), token_cache.txid_vs_tokenid.len());
```

Comment:
Now `token_cache` and `txid_vs_tokenid` are flushed in separate loops.

Also `txid_vs_tokenid` stores operations instead of plain values, because this information is required to properly implement `TransactionVerifierStorage::del_token_id`

Reply:
Btw, please make this a debug_assert.